WordPress/gutenberg
# Font Collections: update registration function signature and add caching

## What?

- Updates `wp_register_font_collection` so that it takes a `$slug` and optional `$arg` array, but does not handle JSON
- Adds `wp_register_font_collection_from_json` function specifically for registering a collection from a JSON file
- Adds static caching for accessing all font collections plus transient caching for remote collections
- Removes `WP_Font_Collection::get_content` in favor of accessing all properties directly (e.g. `$collection->font_families`)

## Why?

- Simplifies/standardizes how collection properties are accessed
- Improves performance for accessing font collections from json with caching
- Brings PHP function signatures to be more similar to other Core `register_` functions

## Testing Instructions

- Ensure the google fonts collection works as expected (with modified json file [from this PR](https://github.com/WordPress/google-fonts-to-wordpress-collection/pull/15) ).
- Try registering a font collection in PHP, and make sure it works as expected

Example PHP font collection for testing:

In a plugin:

```php
wp_register_font_collection_from_json( __DIR__ . '/test-font-collection.json' );
```

`test-font-collection.json`

```json
{
	"$schema": "https://schemas.wp.org/trunk/font-collection.json",
	"slug": "test-font-collection",
	"name": "Test Font Collection",
	"description": "A test font collection.",
	"font_families": [
		{
			"font_family_settings": {
				"name": "Open Sans",
				"fontFamily": "Open Sans, sans-serif",
				"slug": "open-sans",
				"fontFace": [
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memSYaGs126MiZpBA-UvWbX2vVnXBbObj2OVZyOOSr4dVJWUgsiH0C4iY1M2xLER.woff2",
						"fontWeight": "300",
						"fontStyle": "normal",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-300-normal.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memSYaGs126MiZpBA-UvWbX2vVnXBbObj2OVZyOOSr4dVJWUgsjZ0C4iY1M2xLER.woff2",
						"fontWeight": "400",
						"fontStyle": "normal",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-400-normal.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memSYaGs126MiZpBA-UvWbX2vVnXBbObj2OVZyOOSr4dVJWUgsjr0C4iY1M2xLER.woff2",
						"fontWeight": "500",
						"fontStyle": "normal",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-500-normal.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memSYaGs126MiZpBA-UvWbX2vVnXBbObj2OVZyOOSr4dVJWUgsgH1y4iY1M2xLER.woff2",
						"fontWeight": "600",
						"fontStyle": "normal",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-600-normal.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memSYaGs126MiZpBA-UvWbX2vVnXBbObj2OVZyOOSr4dVJWUgsg-1y4iY1M2xLER.woff2",
						"fontWeight": "700",
						"fontStyle": "normal",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-700-normal.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memSYaGs126MiZpBA-UvWbX2vVnXBbObj2OVZyOOSr4dVJWUgshZ1y4iY1M2xLER.woff2",
						"fontWeight": "800",
						"fontStyle": "normal",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-800-normal.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memQYaGs126MiZpBA-UFUIcVXSCEkx2cmqvXlWq8tWZ0Pw86hd0Rk5hkaVIUwaERZjA.woff2",
						"fontWeight": "300",
						"fontStyle": "italic",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-300-italic.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memQYaGs126MiZpBA-UFUIcVXSCEkx2cmqvXlWq8tWZ0Pw86hd0Rk8ZkaVIUwaERZjA.woff2",
						"fontWeight": "400",
						"fontStyle": "italic",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-400-italic.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memQYaGs126MiZpBA-UFUIcVXSCEkx2cmqvXlWq8tWZ0Pw86hd0Rk_RkaVIUwaERZjA.woff2",
						"fontWeight": "500",
						"fontStyle": "italic",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-500-italic.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memQYaGs126MiZpBA-UFUIcVXSCEkx2cmqvXlWq8tWZ0Pw86hd0RkxhjaVIUwaERZjA.woff2",
						"fontWeight": "600",
						"fontStyle": "italic",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-600-italic.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memQYaGs126MiZpBA-UFUIcVXSCEkx2cmqvXlWq8tWZ0Pw86hd0RkyFjaVIUwaERZjA.woff2",
						"fontWeight": "700",
						"fontStyle": "italic",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-700-italic.svg"
					},
					{
						"src": "https://fonts.gstatic.com/s/opensans/v40/memQYaGs126MiZpBA-UFUIcVXSCEkx2cmqvXlWq8tWZ0Pw86hd0Rk0ZjaVIUwaERZjA.woff2",
						"fontWeight": "800",
						"fontStyle": "italic",
						"fontFamily": "Open Sans",
						"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans-800-italic.svg"
					}
				],
				"preview": "https://s.w.org/images/fonts/17.6/previews/open-sans/open-sans.svg"
			},
			"categories": ["sans-serif"]
		}
	],
	"categories": [
		{
			"name": "Sans Serif",
			"slug": "sans-serif"
		}
	]
}
```
